### DISCUSSION TITLE
📝 Nova estrutura de organização de termos do dicionário

### DISCUSSION THREAD

**Discussion by andrepg** [levxyca/diciotech#171] [Ideas]
Bom, seguindo a nova implementação da versão 2, uma das melhorias a se abrir discussão é aquela relacionada à estrutura dos ~verbetes~ termos (_preciso me acostumar ainda, sorry_). Fiz um pequeno estudo por agora, e tenho algumas observações a listas antes de propor mudanças.

* Temos, até o momento, 136 termos cadastrados
* Ainda não temos uma palavra para cada letra, ou seja: há letras não catalogadas ainda
* Os termos estão todos localizados num arquivo JSON apenas
* A letra com mais termos tem menos de 15, à data de escrita desta proposição

> O [dashboard do Looker Studio](https://lookerstudio.google.com/reporting/163e863f-29e6-4042-939c-68740185df82) está disponível para visualização. Por enquanto limitarei a edição a quem precisar.

> Os termos acentuados estão aparecendo separados de suas respectivas letras no Looker. Creio que seja apenas uma questão de transformar o parâmetro no painel. Mas é só um aviso.

## 👀 Considerações

* A estrutura de pastas foi pensada com a possibilidade de internacionalização do projeto.
* Minimizar conflitos ao adicionar ou atualizar termos é uma premissa básica nesta tentativa.
* Organização dos arquivos visa implementar uma lógica básica de dicionário, seguindo a filosofia do projeto

### 📁 Estrutura de pasta
```
+ src/
 \_ dict/
     \ -- [lang]
	   \-- index.js
       |--- [a, b, c].json
```

### 💬 Sobre as pastas

* Uma pasta `dictionaries`, contendo pastas nomeados conforme a linguagem (`pt-BR`, `en-US`, etc.).
* A pasta `dictionaries` contém uma pasta para cada linguagem disponível — `pt-BR`, `en-US`, `es-ES`, etc.

Cada pasta de linguagem contém um _index.js_, fazendo uma referência do tipo chave/valor por letra e arquivo. A definição de como importar pode ser planejada durante o desenvolvimento. A proposição de separar um arquivo para cada letra já foi sugerida no issue #116 durante discussão de recurso. 

```javascript
// index.js
import termsA from "./a.json";
...
import termsZ from "./z.json";

const terms = {
  a: termsA,
  b: termsB,
  ...
  z: termsZ
}
```

Dessa forma podemos garantir uma futura atualização de outras linguagens, adicionando outras pastas correspondentes. A implementação da localização fica por conta do código Javascript no backend.

Acredito que esta estrutura facilitaria o filtro por letras e ordenação personalizada no futuro. A proposta fica, então, aberta a discussão para a comunidade.

**Comment by george-gca** [comment]
Poderia inclusive fazer um loop por todos os arquivos json na pasta e importar, ao invés de manualmente fazer um `import termsA from "./a.json";` pra cada letra. Fica menor o código e menos passível de alguém esquecer de importar algo, já que dentro dessa pasta os json vão ser exclusivamente de dados dos termos mesmo.

**Reply by andrepg** [reply]
Genial! 🤯

Fiquei no receio de sugerir desta forma por não estar 100% familiarizado com o código JS. Honestamente é uma excelente ideia “automatizar” a importação de arquivos, uma possibilidade também contemplada quando da sugestão de padrões de nomenclatura por linguagem e letra. 

Desta forma, ainda, deveria ser possível eliminar a necessidade de um index.js por pasta de linguagem e levá-lo para a pasta `[lang]` em si, fazendo a importação da linguagem correta de uma só vez; 

Eu só não concebi, ainda, como nomear estes _imports_ para declaração (e por isso abrimos discussão 😄)

**Comment by george-gca** [comment]
Outra coisa, os termos internamente poderiam estar em ordem alfabética por chave.

**Comment by george-gca** [comment]
O que acham de depois que a #140 for "mergeada" na main, voltarmos nesse ponto? Acho que seria o momento perfeito pra adicionar essas mudanças. Lembrando que como o @andrepg disse:

> Minimizar conflitos ao adicionar ou atualizar termos é uma premissa básica nesta tentativa.

Isso ia facilitar a adição de novos termos, visto que cada novo termo atualmente mexe sempre no mesmo arquivo.

**Reply by andrepg** [reply]
Sim @george-gca, acredito que podemos voltar. 

Estive pensando melhor sobre a estrutura e queria discutir um pouco mais sobre a divisão dos arquivos. Vocês acreditam ser viável a manutenção de um arquivo por termo, em lugar de um por letra, e uma importação "automática" ? 

Temos o PR #112 para introduzir um CLI, poderíamos usar essa modificação para pavimentar um caminho mais fácil na geração de novos termos com o terminal - já que seria um arquivo novo em lugar de uma modificação, mais fácil de lidar no código. 

O que te parece, @levxyca ?

**Reply by george-gca** [reply]
Não sei se concordo. Atualmente tem uns 150 termos, e a tendência é aumentar com o tempo. Mesmo que sejam arquivos pequenos, seriam muitos. Acho que dividindo por letra já ajuda um pouco a evitar os merges.

Outra possibilidade também seria alterar a forma do arquivo, por exemplo pra um yml. Fiz um teste rápido e o json minificado tem 69,636 KB enquanto o yml tem 71,311 KB. Mas não sei o quanto isso iria melhorar o merge, se o algoritmo de merge ia conseguir fazer mais merges automáticos.

**Reply by levxyca** [reply]
Concordo, também acho que muitos arquivos vai acabar atrapalhando, mas é importante termos uma melhor forma de organizarmos esses termos, atualmente até pra mim ás fica difícil buscar se um termo já existe ou não 😅

Gosto da ideia de dividir por letra 🤔

**Reply by george-gca** [reply]
Dividir por letra e ordenar por ordem alfabética pela chave seria legal. E acho que dá até pra fazer o validador de json validar isso também, tem só que ver se dá pra criar um json schema pra isso.

**Reply by andrepg** [reply]
Olá, colegas.

Estou sumido por razões de _deadline_ **muito** próximo. Mas continuo pensando nesta implementação. Uma coisa tem me chamado atenção, @levxyca.

Temos o #12 para redesenhar o site, porém está parado pelo Hacktober. Vamos voltar com ele, há possibilidade de isto nos afetar de alguma forma? 

Logo após a resposta aqui vou baixar o projeto novamente e configurar no meu VSCode.

**Reply by george-gca** [reply]
Na verdade ele tinha parado pelo hacktoberfest do ano passado, então acho que podemos considerar ele abandonado haha.

**Reply by andrepg** [reply]
Perfeito. Guardemos esse trabalho porque fiquei fã dessa UI pra gente implementar 😄

**Reply by levxyca** [reply]
Exatamente isso, acabou que não houve tempo pra voltar pra essa implementação, mas a issue ainda tá lá registrada pois é uma UI bonitona que eu quero muito que a gente implemente um dia!
